roxctl/common/k8s_port_forward.go

```diff
@@ -0,0 +1,163 @@
+
+	if config.APIPath == "" {
+		config.APIPath = "/api"
+	}
+	if config.NegotiatedSerializer == nil {
+		// This codec factory ensures the resources are not converted. Therefore, resources
+		// will not be round-tripped through internal versions. Defaulting does not happen
+		// on the client.
+		config.NegotiatedSerializer = scheme.Codecs.WithoutConversion()
+	}
+	if err := restclient.SetKubernetesDefaults(config); err != nil {
+		return errPortForwarding.CausedBy(err)
+	}
+	return nil
+}
+
+func runPortForward() (uint16, error) {
+	getter := genericclioptions.NewConfigFlags(true)
+	c := getter.ToRawKubeConfigLoader()
+	restConfig, err := c.ClientConfig()
+	if err != nil {
+		return 0, errPortForwarding.CausedBy(err)
+	}
+	namespace, _, err := c.Namespace()
+	if err != nil {
+		return 0, errPortForwarding.CausedBy(err)
+	}
+
+	if err := setKubernetesDefaults(restConfig); err != nil {
+		return 0, errPortForwarding.CausedBy(err)
+	}
+	coreClient, err := corev1client.NewForConfig(restConfig)
+	if err != nil {
+		return 0, errPortForwarding.CausedBy(err)
+	}
+	podClient, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return 0, errPortForwarding.CausedBy(err)
+	}
+
+	sortBy := func(pods []*corev1.Pod) sort.Interface { return sort.Reverse(podutils.ActivePods(pods)) }
+	forwardablePod, _, err := polymorphichelpers.GetFirstPod(coreClient, namespace, "app=central", 10*time.Second, sortBy)
+	if err != nil {
+		return 0, errPortForwarding.CausedBy(err)
+	}
+
+	pod, err := podClient.CoreV1().Pods(namespace).Get(context.TODO(), forwardablePod.Name, metav1.GetOptions{})
+	if err != nil {
+		return 0, errPortForwarding.CausedBy(err)
+	}
+	if pod.Status.Phase != corev1.PodRunning {
+		return 0, fmt.Errorf("unable to forward port because pod is not running. Current status=%v", pod.Status.Phase)
+	}
+
+	client, err := restclient.RESTClientFor(restConfig)
+	if err != nil {
+		return 0, errPortForwarding.CausedBy(err)
+	}
+
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, os.Interrupt)
+	defer signal.Stop(signals)
+
+	stopChannel := make(chan struct{}, 1)
+	readyChannel := make(chan struct{})
+
+	go func() {
+		<-signals
+		if stopChannel != nil {
+			close(stopChannel)
+		}
+	}()
+
+	req := client.Post().
+		Resource("pods").
+		Namespace(namespace).
+		Name(pod.Name).
+		SubResource("portforward")
+
+	transport, upgrader, err := spdy.RoundTripperFor(restConfig)
+	if err != nil {
+		return 0, errPortForwarding.CausedBy(err)
+	}
+	streams := genericclioptions.IOStreams{}
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: transport}, "POST", req.URL())
+
+	forwarder, err := portforward.New(dialer, []string{"0:8443"}, stopChannel, readyChannel, streams.Out, streams.ErrOut)
+	if err != nil {
+		return 0, errPortForwarding.CausedBy(err)
+	}
+
+	errChan := make(chan error)
+	go func() {
+		errChan <- forwarder.ForwardPorts()
+	}()
+	select {
+	case <-readyChannel:
```

Comment:
Do we need to wait on `stopChannel` here too?

Reply:
Probably not as stopChannel will cause `ForwardPorts` to return `nil`

Reply:
No, we don't need to wait for `stopChannel`. This is to shutdown the port forwarding on a signal.